project-minimization-plugin/src/main/kotlin/org/plan/research/minimization/plugin/execution/comparable/StacktraceExceptionComparator.kt

```diff
@@ -0,0 +1,84 @@
+package org.plan.research.minimization.plugin.execution.comparable
+
+import org.plan.research.minimization.plugin.execution.IdeaCompilationException
+import org.plan.research.minimization.plugin.execution.exception.KotlincException
+import org.plan.research.minimization.plugin.logging.statLogger
+import org.plan.research.minimization.plugin.model.exception.CompilationException
+import org.plan.research.minimization.plugin.model.exception.ExceptionComparator
+
+import org.apache.commons.text.similarity.JaccardSimilarity
+import org.apache.commons.text.similarity.LevenshteinDistance
+
+import kotlin.math.exp
+import kotlin.math.max
+
+class StacktraceExceptionComparator(
+    private val generalComparator: ExceptionComparator,
+) : ExceptionComparator {
+    override fun areEquals(exception1: CompilationException, exception2: CompilationException): Boolean {
+        if (exception1 is IdeaCompilationException && exception2 is IdeaCompilationException) {
+            return exception1.kotlincExceptions.zip(exception2.kotlincExceptions).all {pair -> areEquals(pair.first, pair.second) }
+        }
+
+        if (exception1 !is KotlincException || exception2 !is KotlincException) {
+            return generalComparator.areEquals(exception1, exception2)
+        }
+
+        val stacktrace1 = getStacktrace(exception1) ?: return generalComparator.areEquals(exception1, exception2)
+        val stacktrace2 = getStacktrace(exception2) ?: return generalComparator.areEquals(exception1, exception2)
+
+        return isSimilar(stacktrace1, stacktrace2)
+    }
+
+    private fun getStacktrace(exception: KotlincException): List<String>? = when (exception) {
+        is KotlincException.BackendCompilerException -> parseMessage(exception.additionalMessage) + parseStacktrace(exception.stacktrace)
+        is KotlincException.GenericInternalCompilerException -> parseMessage(exception.message) + parseStacktrace(exception.stacktrace)
+        is KotlincException.GeneralKotlincException -> null
+        is KotlincException.KspException -> parseMessage(exception.message) + parseStacktrace(exception.stacktrace)
+    }
+
+    // leave informative part of stacktrace lines
+    private fun parseStacktrace(stacktrace: String): List<String> = stacktrace.lines()
+        .map { it.trim() }
+        .takeWhile { it.startsWith("at org.jetbrains.kotlin.") }
+        .map { it.removePrefix("at org.jetbrains.kotlin.").trim() }
+
+    private fun parseMessage(message: String?): List<String> =
+        message?.let { listOf(it) } ?: emptyList()
+
+    private fun isSimilar(stackList1: List<String>, stackList2: List<String>): Boolean {
+        var similarity: Double = ABSOLUTE_SIMILARITY
+
+        stackList1.zip(stackList2).forEachIndexed { index, (line1, line2) ->
```

Comment:
What if sizes are different?

Reply:
We discussed it at meeting and decided to ignore such case, as anyway the most important info located in a few first lines, so we can just ignore the rest